File: Add 1 to a number represented as linked list - GFG/add-1-to-a-number-represented-as-linked-list.cpp
// { Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h> 
using namespace std; 

struct Node
{
    int data;
    struct Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
};

void printList(Node* node) 
{ 
    while (node != NULL) { 
        cout << node->data; 
        node = node->next; 
    }  
    cout<<"\n";
} 


 // } Driver Code Ends
//User function template for C++

/* 

struct Node
{
    int data;
    struct Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
};

*/

class Solution
{
    public:
    int c=0;
    Node*help(Node*head)
    {
        if(!head->next)
        {
            head->data=head->data+1;
            c=head->data/10;
            head->data=head->data%10;
            return head;
        }
        help(head->next);
        head->data=head->data+c;
        c=head->data/10;
        head->data=head->data%10;
        return head;
    }
    Node* addOne(Node *head) 
    {
        if(!head)
        return head;
        // if(!head->next)
        // {
        //     head->data+=1;
        //     return head;
        // }
        head=help(head);
        if(c!=0)
        {
            Node*newHead=new Node(c);
            newHead->next=head;
            return newHead;
        }
        return head;
    }
};

// { Driver Code Starts.

int main() 
{ 
    int t;
    cin>>t;
    while(t--)
    {
        string s;
        cin>>s;
        
        Node* head = new Node( s[0]-'0' );
        Node* tail = head;
        for(int i=1; i<s.size(); i++)
        {
            tail->next = new Node( s[i]-'0' );
            tail = tail->next;
        }
        Solution ob;
        head = ob.addOne(head);
        printList(head); 
    }
    return 0; 
} 
  // } Driver Code Ends

File: Add 1 to a number represented as linked list - GFG/add-1-to-a-number-represented-as-linked-list.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;
class Node
{
    int data;
    Node next;
    
    Node(int x)
    {
        data = x;
        next = null;
    }
}
class GfG
{
    public static void printList(Node node) 
    { 
        while (node != null)
        { 
            System.out.print(node.data);
            node = node.next; 
        }  
        System.out.println();
    } 
    public static void main(String args[])throws IOException
        {
            Scanner sc = new Scanner(System.in);
            int t = sc.nextInt();
            while(t-->0)
                {
                    String s = sc.next();
                    Node head = new Node( s.charAt(0) - '0' );
                    Node tail = head;
                    for(int i=1; i<s.length(); i++)
                    {
                        tail.next = new Node( s.charAt(i) - '0' );
                        tail = tail.next;
                    }
                    Solution obj = new Solution();
                    head = obj.addOne(head);
                    printList(head); 
                }
        }
}
// } Driver Code Ends


/*
class Node{
    int data;
    Node next;
    
    Node(int x){
        data = x;
        next = null;
    }
} 
*/

class Solution
{
    static int carry=0;
    static Node newHead;
    public static Node fun(Node head)
    {
        if(head==null)
        return null;
        if(head.next==null)
        {
            head.data+=carry+1;
            carry=head.data/10;
            head.data%=10;
            return head;
        }
        head.next=fun(head.next);
        head.data+=carry;
        carry=head.data/10;
        head.data%=10;
        return head;
    }
    public static Node addOne(Node head) 
    { 
        carry=0;
        newHead=fun(head);
        if(carry!=0)
        {
            Node temp=new Node(carry);
            temp.next=newHead;
            newHead=temp;
        }
        return newHead;
        
    }
}
